_posts/2023-05-02-Rotating-Planes-on-VKMS.md

```diff
@@ -0,0 +1,143 @@
+Each line is represented by a staging buffer, which contains the composition for one plane, and an output buffer, which contains the composition of all planes in z-pos order.
+For each line, we query an array by the first pixel of the line and transpose the whole array linearly, performing the proper pixel conversion.
+The composition of the line can be summarized by:
+
+```c
+void vkms_compose_row(struct line_buffer *stage_buffer, struct vkms_plane_state *plane, int y)
+{
+	struct pixel_argb_u16 *out_pixels = stage_buffer->pixels;
+	struct vkms_frame_info *frame_info = plane->frame_info;
+	u8 *src_pixels = get_packed_src_addr(frame_info, y);
+	int limit = min_t(size_t, drm_rect_width(&frame_info->dst), stage_buffer->n_pixels);
+ 
+	for (size_t x = 0; x < limit; x++, src_pixels += frame_info->cpp)
+		plane->pixel_read(src_pixels, &out_pixels[x]);
+}
+```
+
+With that in mind, we can think that rotating a plane is a matter of changing how we read and interpret the lines.
+Let's think about the `reflect-x` operation. 
+
+```
+		|*********|$$$$$$$$$|                |$$$$$$$$$|*********|
+		|*********|$$$$$$$$$|  ----------->  |$$$$$$$$$|*********|
+		|#########|@@@@@@@@@|   reflect-x    |@@@@@@@@@|#########|
+		|#########|@@@@@@@@@|                |@@@@@@@@@|#########|
+```
+
+Thinking that the VKMS composition happens line-by-line, we just need to read the line in reverse order.
+So, instead of start reading the pixels from left to right, we need to start reading the pixels from right to left.
+We can implement this in a very simple way by getting the limit of the line and subtracting the current `x` position:
+
+```c
+static int get_x_position(const struct vkms_frame_info *frame_info, int limit, int x)
+{
+	if (frame_info->rotation & DRM_MODE_REFLECT_X)
+		return limit - x - 1;
+	return x;
+}
+```
+
+Similarly, for the `reflect-y` operation, we need to start reading the plane from the last line, instead of reading it from the first line.
+
+```
+		|*********|$$$$$$$$$|                |#########|@@@@@@@@@|
+		|*********|$$$$$$$$$|  ----------->  |#########|@@@@@@@@@|
+		|#########|@@@@@@@@@|   reflect-y    |*********|$$$$$$$$$|
+		|#########|@@@@@@@@@|                |*********|$$$$$$$$$|
+```
+
+This can be simply performed by changing the `y` on the external composition loop.
+Similarly from the `reflect-x` case, we can get the `y` limit and subtract the current `y` position.
+
+```c
+static int get_y_pos(struct vkms_frame_info *frame_info, int y)
+{
+	if (frame_info->rotation & DRM_MODE_REFLECT_Y)
+		return drm_rect_height(&frame_info->rotated) - y - 1;
+	return y;
+}
+```
+
+So, in order to implement the rotation in VKMS, we just need to change how we interpret the boundaries of the plane and read accordingly. 
+
+This might seem a bit odd because we could just rotate the `src` rectangle by using `drm_rect_rotate`, but this wouldn't work as the composition in VKMS is performed line-by-line and the pixels are accessed linearly.
```

Comment:
```suggestion
This might seem odd because we could just rotate the `src` rectangle by using `drm_rect_rotate`, but this wouldn't work as the composition in VKMS is performed line-by-line and the pixels are accessed linearly.
```